src/Shared/FileMatcher.cs

```diff
@@ -1062,10 +1062,27 @@ struct RecursionState
             {
                 return Enumerable.Empty<string>();
             }
+
+            // Back-compat hack: We don't use case-insensitive file enumeration I/O on Linux so the behavior is different depending
+            // on the NeedsToProcessEachFile flag. If the flag is false and matching is done within the _getFileSystemEntries call,
+            // it is case sensitive. If the flag is true and matching is handled with MatchFileRecursionStep, it is case-insensitive.
+            // TODO: Can we fix this by using case-insensitive file I/O on Linux?
+            bool forceFileProcessing = false;
```

Comment:
tiny nit:
If you give this true as a default value, you don't have to reassign it for the first case.

Reply:
Actually, let's remove the local altogether and use the existing `NeedsToProcessEachFile` field on the struct.